src/gadgets/mpt_update.rs

```diff
@@ -543,6 +375,241 @@ fn configure_storage<F: FieldExt>(cb: &mut ConstraintBuilder<F>, config: &MptUpd
+    );
+    cb.assert(
+        "path_type is Common",
+        config.selector.current(),
+        config.path_type.matches(PathType::Common),
+    );
+    cb.add_constraint(
+        "depth is 0",
+        config.selector.current(),
+        config.depth.current(),
+    );
+    for variant in MPTProofType::iter() {
+        let conditional_constraints = |cb: &mut ConstraintBuilder<F>| match variant {
+            MPTProofType::AccountDestructed | MPTProofType::AccountDoesNotExist => todo!(),
+            MPTProofType::BalanceChanged
+            | MPTProofType::CodeHashExists
+            | MPTProofType::CodeSizeExists
+            | MPTProofType::NonceChanged
+            | MPTProofType::PoseidonCodeHashExists
+            | MPTProofType::StorageChanged
+            | MPTProofType::StorageDoesNotExist => cb.add_constraint(
+                "direction is 0",
+                config.selector.current(),
+                config.direction.current(),
+            ),
+        };
+        cb.condition(config.proof_type.matches(variant), conditional_constraints);
+    }
+    // add constraints that sibling = old_path_key and new_path_key
+}
+
+fn configure_account_leaf1<F: FieldExt>(cb: &mut ConstraintBuilder<F>, config: &MptUpdateConfig) {
+    cb.assert(
+        "previous is AccountLeaf0",
+        config.selector.current(),
+        config
+            .segment_type
+            .previous_matches(SegmentType::AccountLeaf0),
+    );
+    cb.assert(
+        "next is AccountLeaf2",
+        config.selector.current(),
+        config.segment_type.next_matches(SegmentType::AccountLeaf2),
+    );
+    cb.assert(
+        "path_type is Common",
+        config.selector.current(),
+        config.path_type.matches(PathType::Common),
+    );
+    cb.add_constraint(
+        "depth is 0",
+        config.selector.current(),
+        config.depth.current(),
+    );
+    for variant in MPTProofType::iter() {
+        let conditional_constraints = |cb: &mut ConstraintBuilder<F>| match variant {
+            MPTProofType::AccountDestructed | MPTProofType::AccountDoesNotExist => todo!(),
+            MPTProofType::BalanceChanged
+            | MPTProofType::CodeHashExists
+            | MPTProofType::CodeSizeExists
+            | MPTProofType::NonceChanged
+            | MPTProofType::StorageChanged
+            | MPTProofType::StorageDoesNotExist => cb.add_constraint(
+                "direction is 0",
+                config.selector.current(),
+                config.direction.current(),
+            ),
+            MPTProofType::PoseidonCodeHashExists => {
+                cb.add_constraint(
+                    "direction is 1",
+                    config.selector.current(),
+                    Query::one() - config.direction.current(),
+                );
+                // TODO: add poseidon lookup
+            }
+        };
+        cb.condition(config.proof_type.matches(variant), conditional_constraints);
```

Comment:
I think having this table of allowed `(SegmentType, ProofType, direction)` tuples will be great for readability of the circuit. I'm not sure if it's better to use (many) custom gates instead?. Custom gates would looks something like this:
```
for (segment, proof_type, direction) in allowed_tuples.iter() {
  cb.condition(config.segment.matches(segment).and(config.proof_type.matches(proof_type)), |cb| {
    cb.require_equal("direction matches segment and proof_type", Query::from(direction), config.direction.current();
  });
}
```

Reply:
for such a custom gate, i think we need ''onehot' gadget for each matching (or the degree would easily exceed limit) and this would cost a bunch of cols?

Reply:
Moved to a new issue https://github.com/scroll-tech/mpt-circuit/issues/35

Reply:
Hi @noel2004 and @z2trillion, I add `SegmentProofLookup` to lookup `[segment_type, proof_type, direction]` tuple (and called in corresponding functions). Please help review when you have time. Thanks.

Reply:
The `SegmentProofConfig` should contain fixed cols. And put all possible combinations on the values of `[segment_type, proof_type, direction]`. You can pick these combinations from the condition gate before.
Like I have mentioned above:

``` rust
// fill these values in assign of SegmentProofConfig
[SegmentType::AccountLeaf1, MPTProofType::BalanceChanged, 0],
[SegmentType::AccountLeaf1, MPTProofType::CodeHashExists, 0],
[SegmentType::AccountLeaf1, MPTProofType::CodeSizeExists, 0],
[SegmentType::AccountLeaf1, MPTProofType::NonceChanged, 0],
[SegmentType::AccountLeaf1, MPTProofType::StorageChanged, 0],
[SegmentType::AccountLeaf1, MPTProofType::StorageDoesNotExist, 0],
[SegmentType::AccountLeaf1, MPTProofType::PoseidonCodeHashExists, 1],
```

Reply:
Add a constant `SEGMENT_PROOF_DIRECTION_TUPLES` of (segment_type, proof_type, direction) tuple values, and update `SegmentProofLookup` assign function in commit [5014eda](https://github.com/scroll-tech/mpt-circuit/pull/32/commits/5014eda786adca4579fa41f57b99adbc3445c92a).

Reply:
Fixed to `FixedColumn` (from `AdviceColumn`) in commit [10919f3](https://github.com/scroll-tech/mpt-circuit/pull/32/commits/10919f3bf6649848994148bfd7d4761d4c9fcfb1).